### ISSUE TITLE
18Scan issue: runTo="tokenonly" for an off-map city is not obeyed.

### ISSUE THREAD

**Issue by erik-vos** [Rails-18xx/Rails#234]
In 18Scan, off-map cities may only be run to if a token of the running company is present there. This is configured by the runTo="tokenonly" attribute in Map.xml. The attached picture suggests that this runTo value (understandably) has not yet been implemented.
![18Scan_IgnoresTokenOnly](https://user-images.githubusercontent.com/13908049/83169985-20cdd680-a114-11ea-925b-4fe72ac76a16.jpg)

As I am not familiar at all with the routing code, I wonder if anyone can do something about it. I will push the current 18Scan code to erik-vos/18Scan. Here follows a save file.

[18Scan_IgnoresTokenOnly.zip](https://github.com/Rails-18xx/Rails/files/4697115/18Scan_IgnoresTokenOnly.zip)

**Comment by erik-vos** [comment]
I have found the cause: the Map.xml defaults are ineffective because they are parsed _after_ being called during MapHex parsing, so these are all set to null. It's unclear to me how this could have happened. Anyway, so far it does not seem to be a route finding problem, that code looks OK.

I am working on it.

**Comment by erik-vos** [comment]
After a lot of debugging and diagram scribbling I think I have reached the bottom of this issue. At configuration time there are more use-before-defined cases than the one I mentioned above. At execution time it works, if I specify tokenonly per MapHex rather than as an offmap default. It even works too well: even token laying is prevented! So I will have to fix that first.

The background of this all is a discussion I had with Stefan during 2011/12, about which I have resurfaced a few emails. At that time I had implemented a new class Access to deal with these Stop properties. Access has been part of later versions of Rails 1.x; all I have kept of those days is a v1.7.12 source and a v1.9.1 jar, both containing Access. In Rails 2.0, Stefan had renamed Access to StopType, a name that confused me at first, because we are not dealing with types but with properties that (in theory) can be set per individual Tile or MapHex. He also had changed the configuration code pretty heavily, with the above results. 

I'm not going to reintroduce the original Access class, but surely the Stop property config code will get closer to my earlier version of it.

**Comment by erik-vos** [comment]
I have fixed the token laying prevention issue mentioned in my previous post, by adding a new boolean parameter called "running" through 6 layers of network code down to Stop.isRunToAllowed(). It is always true (to show the valid train routes at all times), but false in the ORUIManager code to find valid hexes for tile and token laying.

The commit is pushed to erik-vos/18Scan. It also includes small fixes for some harmless 'errors' that IntelliJ irritatingly complained about, at least to me. And there is some debugging code left that I still may need further on.

**Comment by erik-vos** [comment]
First part included in PR #238. Defaults still don't work.

**Comment by erik-vos** [comment]
I have pushed fixes to erik-vos/18Scan. The immediate problem has been fixed now, but the changes were more extensive that I expected, and there is still some code cleanup and bringing more consistency to do, so I won't close the issue now.

Quite some problems with the Rails 2.0 implementation of the Stop properties have surfaced:

- The stop type enums and associated defaults had been defined separately (and in my view redundantly) both in Station and in StopType. 
- The defaults in StopType and in Map.xml did not work because they were invoked at a time that the tiles had not yet been configured, so the type was unknown. As a result, all stop properties were set to null.
- This made the independent Station property defaults finally prevail, which are fixed per tile.
- Loop prevention was based on city/offmap but also on station names, and the latter was done in Tiles.xml, which is a no-no, because these files are created programmatically and can be overwritten.

The whole configuration process was quite intransparent, and I felt I had to return, at least partially, to the original Stop properties framework as I had created for Rails 1.x, and that Stefan had partially built upon in 2.0. 

For those interested, I attach the contents of a 2012 mail about stop properties. I have deprecated loop prevention based on city/station names, will probably also drop the "loop" property, and instead intend to implement the much more versatile mutexId concept. Also, runTo and runThrough have been implemented in the meantime.
[StopFrameworkMail.txt](https://github.com/Rails-18xx/Rails/files/4786783/StopFrameworkMail.txt)

I am also considering to extend the runTo parameter to enable restricting hex access to specific train types, such as the mines in 1837 and 18VA where only coal/goods trains may run to. This would obviate the need for specific Modifier classes, or make these configurable. But that all goes way beyond the topic if this issue.

**Comment by erik-vos** [comment]
I have pushed two commits to erik-vos/18Scan. The first fixes the above problems, and a few more:
- The term "stop type" was used in two different ways:
  - StopType as the name of a class, holding the access parameters and code to parse these. I have back-renamed this class to Access, but it is not identical to the similarly named class of yore (i.e. Rails 1.x).
  - As a enum defining various stop/station types, defined twice in Station (as Type) and in StopType (as Default). I have reduced this to one simple code-less enum Type in Stop, so Stop.Type is now the universal stop/station type. Defaults are now held in simple EnumMap<Stop.Type, Access> objects.
- City names were not just nice extra information, but also used in two ways:
  - As a way to identify a certain location in a Modifier. This was used in 18Kaas ('Ruhrgebied'). Perhaps this could have been done in another way, but I have retained the possibility of this usage.
  - As a way to identify hex(es) where loop restrictions apply. I have partially dropped this usage of city names, replacing it by mutexId as described in the attachment in the previous post. The loop variable has also been dropped, Only for off-map areas the city names are still used, filling the mutexId(s) of the relevant hex(es) if no mutexId is configured. As far as I know all games except 18PA do not allow trains to access the same area twice, so this is a good default. 18VA/Map.xml shows how the mutexId can be used to allow trains running between the three different off-map area types of that game (off-map cities, ports, coal mine disctricts), but not from and to the same type.

I believe that the new code is much more transparent and simpler to understand (at least it is for me).

The second commit fixes a problem in 18EU, where the Berlin green upgrade did not create a third stop for the extra station. This had been worked around by manually adding an invisible third station to the preprinted tile in Tiles.xml, but that file must not be edited as it is generated programatically. This was a pretty easy fix.

 All tests pass. A PR will be created shortly, but before that I want to do some testing and cleanup.

**Comment by erik-vos** [comment]
BTW I have checked the following games for loop restrictions in on-map cities. For only the cities mentioned below I have found such a restriction, which is now configured by specifying a mutexId:
18EU: Paris, Berlin, Vienna.
1835: Berlin.
1837: Vienna (only).
1830, 1826, 18VA, 18Scan, 18AL, 1841, 1851, 1856, SOH: none.
These are the games that I am (somewhat) familiar with, I'm leaving the remaining ones to others.
